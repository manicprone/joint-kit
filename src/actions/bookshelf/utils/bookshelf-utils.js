import objectUtils from '../../../utils/object-utils'
import stringUtils from '../../../utils/string-utils'
import ACTION from '../../../core/constants/action-constants'

const debugLoadDirect = false

// -----------------------------------------------------------------------------
// Accepts the "orderBy" API field value and returns
// the Bookshelf-compatible specification for an order-by clause.
// -----------------------------------------------------------------------------
// The returned specification is an array in the format:
// [{ col: 'updated_at', order: 'desc' }, { col: 'title', order: 'asc' }]
// -----------------------------------------------------------------------------
export function buildOrderBy (fieldValue) {
  const orderBy = []

  if (fieldValue) {
    const columns = fieldValue.split(',')
    for (let i = 0; i < columns.length; i++) {
      let col = columns[i].trim()
      let order = 'asc'
      if (col.length > 0) {
        // Interpret a negative value as a descending option...
        if (col.startsWith('-')) {
          col = col.substr(1)
          order = 'desc'
        }
        orderBy.push({ col, order })
      }
    } // end-for
  } // end-if (fieldValue)

  return orderBy
}

// -----------------------------------------------------------------------------
// Performs the "loadDirect" logic, hoisting relation data to the base
// attributes of the provided Bookshelf data.
// -----------------------------------------------------------------------------
// NOTE: This function mutates the data provided. There is no return value.
// -----------------------------------------------------------------------------
export function loadRelationsToItemBase (itemData, loadDirect = {}, keepAsRelations = []) {
  if (loadDirect.associations) {
    // Loop through all loadDirect requests, moving the specified column data to the item's base attributes...
    loadDirect.associations.forEach((relationName) => {
      const colNames = loadDirect.colMappings[relationName]
      const relationData = (itemData.relations[relationName]) ? itemData.relations[relationName] : null
      let loadDirectData = null

      if (relationData) {
        // Handle collection (many relation)...
        if (relationData.models) {
          loadDirectData = []
          relationData.models.forEach((modelData) => {
            // Multiple, explicit fields...
            if (Array.isArray(colNames)) {
              const colDataSet = {}
              colNames.forEach((colName) => {
                colDataSet[colName] = modelData.attributes[colName]
              })
              loadDirectData.push(colDataSet)
            // Wildcard fields...
            } else if (colNames === '*') {
              loadDirectData.push(modelData.attributes)
            // Single field...
            } else {
              loadDirectData.push(modelData.attributes[colNames])
            }
          })

        // Handle item (1-1 relation)...
        } else {
          /* eslint-disable no-lonely-if */
          // Multiple, explicit fields...
          if (Array.isArray(colNames)) {
            loadDirectData = {}
            colNames.forEach((colName) => {
              loadDirectData[colName] = relationData.attributes[colName]
            })
          // Wildcard fields...
          } else if (colNames === '*') {
            loadDirectData = relationData.attributes
          // Single field...
          } else {
            loadDirectData = relationData.attributes[colNames]
          }
          /* eslint-enable no-lonely-if */
        }
      } // end-if (relationData)

      if (debugLoadDirect) console.log(`[JOINT] [bookshelf-utils:loadRelationsToItemBase] load direct: ${relationName}:${colNames} => ${loadDirectData}`)

      // Copy the column value to a base attribute (using the relation name as the property name)...
      if (loadDirectData) {
        const attrName = stringUtils.toSnakeCase(relationName)
        itemData.attributes[attrName] = loadDirectData
      }

      // If not included in the standard relations, remove the relation data from the item...
      if (!objectUtils.includes(keepAsRelations, relationName)) {
        delete itemData.relations[relationName]
      }
    })
  } // end-if (loadDirect.associations)
}

// -----------------------------------------------------------------------------
// Append a where query to an existing query builder, respecting the type of
// field value and its matchStrategy.
// -----------------------------------------------------------------------------
export function appendWhereClause (queryBuilder, fieldName, value, matchStrategy) {
  switch (matchStrategy) {
    case ACTION.INPUT_FIELD_MATCHING_STRATEGY_EXACT:
      if (Array.isArray(value)) queryBuilder.where(fieldName, 'IN', value)
      else queryBuilder.where(fieldName, '=', value)
      break
    case ACTION.INPUT_FIELD_MATCHING_STRATEGY_CONTAINS:
      // Case insensitive LIKE query
      // Note that case-sensitivity of LIKE differs per DBMS, comparing always
      // with lowercase is potentially slower but more portable.
      queryBuilder.whereRaw('LOWER( ?? ) LIKE ?', [fieldName, `%${value.toLowerCase()}%`])
      break
    default:
      throw new Error(`Unrecognized match strategy "${matchStrategy}"`)
  }
}
