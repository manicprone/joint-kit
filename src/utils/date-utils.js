import moment from 'moment'

function now () {
  return moment.utc().format()
}

export default {
  now
}
