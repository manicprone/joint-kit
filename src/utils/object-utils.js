import lodashObject from 'lodash/object'
import lodashLang from 'lodash/lang'
import lodashArray from 'lodash/array'
import lodashCollection from 'lodash/collection'

export default {
  get: lodashObject.get,
  has: lodashObject.has,
  isEmpty: lodashLang.isEmpty,
  cloneDeep: lodashLang.cloneDeep,
  union: lodashArray.union,
  includes: lodashCollection.includes
}
