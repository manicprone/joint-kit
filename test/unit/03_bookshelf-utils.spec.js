import { beforeEach, describe, expect, it } from 'vitest'
import * as BookshelfUtils from '../../src/actions/bookshelf/utils/bookshelf-utils'

const itemData = {}

// -----------------------
// LIBRARY bookshelf-utils
// -----------------------
describe('BOOKSHELF-UTILS', function () {
  // ---------------------
  // Testing: buildOrderBy
  // ---------------------
  describe('buildOrderBy', function () {
    it('should return an empty array if no value is provided', function () {
      const orderBy = BookshelfUtils.buildOrderBy()

      expect(orderBy).toEqual([])
    })

    it('should return the Bookshelf-compatible spec for a single order value (positive/ascending)', function () {
      const fieldValue = 'title'
      const orderBy = BookshelfUtils.buildOrderBy(fieldValue)

      expect(orderBy).toMatchInlineSnapshot(`
        [
          {
            "col": "title",
            "order": "asc",
          },
        ]
      `)
    })

    it('should return the Bookshelf-compatible spec for a single order value (negative/descending)', function () {
      const fieldValue = '-title'
      const orderBy = BookshelfUtils.buildOrderBy(fieldValue)

      expect(orderBy).toMatchInlineSnapshot(`
        [
          {
            "col": "title",
            "order": "desc",
          },
        ]
      `)
    })

    it('should return the Bookshelf-compatible spec for multiple values (comma-delimited)', function () {
      const fieldValue = '-title,updated_at,status_id'
      const orderBy = BookshelfUtils.buildOrderBy(fieldValue)

      expect(orderBy).toMatchInlineSnapshot(`
        [
          {
            "col": "title",
            "order": "desc",
          },
          {
            "col": "updated_at",
            "order": "asc",
          },
          {
            "col": "status_id",
            "order": "asc",
          },
        ]
      `)
    })

    it('should handle extraneous spaces between values', function () {
      const fieldValue = ' -title, updated_at  ,    status_id'
      const orderBy = BookshelfUtils.buildOrderBy(fieldValue)

      expect(orderBy).toMatchInlineSnapshot(`
        [
          {
            "col": "title",
            "order": "desc",
          },
          {
            "col": "updated_at",
            "order": "asc",
          },
          {
            "col": "status_id",
            "order": "asc",
          },
        ]
      `)
    })

    it('should handle empty values within commas', function () {
      const fieldValue = 'title,  ,   -updated_at  , ,'
      const orderBy = BookshelfUtils.buildOrderBy(fieldValue)

      expect(orderBy).toMatchInlineSnapshot(`
        [
          {
            "col": "title",
            "order": "asc",
          },
          {
            "col": "updated_at",
            "order": "desc",
          },
        ]
      `)
    })
  }) // END - buildOrderBy

  // --------------------------------
  // Testing: loadRelationsToItemBase
  // --------------------------------
  describe('loadRelationsToItemBase', function () {
    beforeEach(() => {
      itemData.attributes = {
        id: 1,
        user_id: 4,
        name: 'Project Apathy',
        status_code: 5
      }
      itemData.relations = {
        profile: {
          id: 9999,
          attributes: {
            id: 1,
            user_id: 4,
            title: 'Functional Fanatic',
            slug: 'functional-fanatic',
            tagline: 'I don\'t have habits, I have algorithms.',
            description: null,
            is_default: true,
            is_live: false
          }
        },
        user: {
          id: 4,
          attributes: {
            id: 4,
            username: 'the_manic_edge',
            external_id: '304',
            email: 'the-manic-edge@demo.com',
            display_name: 'The Manic Edge'
          }
        },
        team: {
          id: 27,
          attributes: {
            id: 27,
            name: 'The Coalition',
            slug: 'the-coalition',
            email: 'team@the-coalition.org',
            member_count: 5
          }
        },
        techConceptTags: {
          length: 5,
          models: [
            { id: 1, attributes: { id: 1, label: 'Software Architecture', key: 'software-architecture', created_by: 1 } },
            { id: 2, attributes: { id: 2, label: 'Machine Learning', key: 'machine-learning', created_by: 1 } },
            { id: 3, attributes: { id: 3, label: 'Blockchain', key: 'blockchain', created_by: 1 } },
            { id: 4, attributes: { id: 4, label: 'Crypto Currency', key: 'crypto-currency', created_by: 8 } },
            { id: 5, attributes: { id: 5, label: 'Big Data', key: 'big-data', created_by: 9 } }
          ]
        },
        codingLanguageTags: {
          length: 3,
          models: [
            { id: 1, attributes: { id: 1, label: 'Java', key: 'java' } },
            { id: 2, attributes: { id: 2, label: '.NET', key: 'dot-net' } },
            { id: 3, attributes: { id: 3, label: 'JavaScript', key: 'javascript' } }
          ]
        },
        softwareTags: {
          length: 3,
          models: [
            { id: 1, attributes: { id: 1, label: 'Vue', key: 'vue', created_by: 1 } },
            { id: 2, attributes: { id: 2, label: 'React', key: 'react', created_by: 4 } },
            { id: 3, attributes: { id: 3, label: 'Express', key: 'express', created_by: 7 } }
          ]
        }
      }
    })

    it('should do nothing if the parsed "loadDirect" info does not contain an "associations" property', function () {
      const originalItemData = { ...itemData }
      const loadDirect = {}

      BookshelfUtils.loadRelationsToItemBase(itemData, loadDirect)

      expect(itemData).toEqual(originalItemData)
    })

    it('should hoist the specified field data to the base attributes of the main resource', function () {
      const loadDirect = {
        associations: ['techConceptTags', 'codingLanguageTags', 'softwareTags', 'user', 'profile', 'team'],
        colMappings: {
          techConceptTags: 'key', // toMany relation, single col
          codingLanguageTags: ['label', 'key'], // toMany relation, multiple explicit cols
          softwareTags: '*', // toMany relation, wildcard (all) cols
          user: 'username', // toOne relation, single col
          profile: ['title', 'tagline', 'is_live'], // toOne relation, multiple explicit cols
          team: '*' // toOne relation, wildcard (all) cols
        }
      }

      BookshelfUtils.loadRelationsToItemBase(itemData, loadDirect)

      expect(itemData).toMatchSnapshot()
    })

    it('should delete the original relation data, if not explicitly included', function () {
      const keepAsRelations = ['softwareTags', 'profile']
      const loadDirect = {
        associations: ['techConceptTags', 'codingLanguageTags', 'softwareTags', 'user', 'profile', 'team'],
        colMappings: {
          techConceptTags: 'key', // toMany relation, single col
          codingLanguageTags: ['label', 'key'], // toMany relation, multiple explicit cols
          softwareTags: '*', // toMany relation, wildcard (all) cols
          user: 'username', // toOne relation, single col
          profile: ['title', 'tagline', 'is_live'], // toOne relation, multiple explicit cols
          team: '*' // toOne relation, wildcard (all) cols
        }
      }

      BookshelfUtils.loadRelationsToItemBase(itemData, loadDirect, keepAsRelations)

      expect(itemData).toMatchSnapshot()
    })
  }) // END - loadRelationsToItemBase
})
