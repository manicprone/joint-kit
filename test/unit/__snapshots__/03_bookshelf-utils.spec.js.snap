// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`BOOKSHELF-UTILS > loadRelationsToItemBase > should delete the original relation data, if not explicitly included 1`] = `
{
  "attributes": {
    "coding_language_tags": [
      {
        "key": "java",
        "label": "Java",
      },
      {
        "key": "dot-net",
        "label": ".NET",
      },
      {
        "key": "javascript",
        "label": "JavaScript",
      },
    ],
    "id": 1,
    "name": "Project Apathy",
    "profile": {
      "is_live": false,
      "tagline": "I don't have habits, I have algorithms.",
      "title": "Functional Fanatic",
    },
    "software_tags": [
      {
        "created_by": 1,
        "id": 1,
        "key": "vue",
        "label": "Vue",
      },
      {
        "created_by": 4,
        "id": 2,
        "key": "react",
        "label": "React",
      },
      {
        "created_by": 7,
        "id": 3,
        "key": "express",
        "label": "Express",
      },
    ],
    "status_code": 5,
    "team": {
      "email": "team@the-coalition.org",
      "id": 27,
      "member_count": 5,
      "name": "The Coalition",
      "slug": "the-coalition",
    },
    "tech_concept_tags": [
      "software-architecture",
      "machine-learning",
      "blockchain",
      "crypto-currency",
      "big-data",
    ],
    "user": "the_manic_edge",
    "user_id": 4,
  },
  "relations": {
    "profile": {
      "attributes": {
        "description": null,
        "id": 1,
        "is_default": true,
        "is_live": false,
        "slug": "functional-fanatic",
        "tagline": "I don't have habits, I have algorithms.",
        "title": "Functional Fanatic",
        "user_id": 4,
      },
      "id": 9999,
    },
    "softwareTags": {
      "length": 3,
      "models": [
        {
          "attributes": {
            "created_by": 1,
            "id": 1,
            "key": "vue",
            "label": "Vue",
          },
          "id": 1,
        },
        {
          "attributes": {
            "created_by": 4,
            "id": 2,
            "key": "react",
            "label": "React",
          },
          "id": 2,
        },
        {
          "attributes": {
            "created_by": 7,
            "id": 3,
            "key": "express",
            "label": "Express",
          },
          "id": 3,
        },
      ],
    },
  },
}
`;

exports[`BOOKSHELF-UTILS > loadRelationsToItemBase > should hoist the specified field data to the base attributes of the main resource 1`] = `
{
  "attributes": {
    "coding_language_tags": [
      {
        "key": "java",
        "label": "Java",
      },
      {
        "key": "dot-net",
        "label": ".NET",
      },
      {
        "key": "javascript",
        "label": "JavaScript",
      },
    ],
    "id": 1,
    "name": "Project Apathy",
    "profile": {
      "is_live": false,
      "tagline": "I don't have habits, I have algorithms.",
      "title": "Functional Fanatic",
    },
    "software_tags": [
      {
        "created_by": 1,
        "id": 1,
        "key": "vue",
        "label": "Vue",
      },
      {
        "created_by": 4,
        "id": 2,
        "key": "react",
        "label": "React",
      },
      {
        "created_by": 7,
        "id": 3,
        "key": "express",
        "label": "Express",
      },
    ],
    "status_code": 5,
    "team": {
      "email": "team@the-coalition.org",
      "id": 27,
      "member_count": 5,
      "name": "The Coalition",
      "slug": "the-coalition",
    },
    "tech_concept_tags": [
      "software-architecture",
      "machine-learning",
      "blockchain",
      "crypto-currency",
      "big-data",
    ],
    "user": "the_manic_edge",
    "user_id": 4,
  },
  "relations": {},
}
`;
